### ISSUE TITLE
gql tag

### ISSUE THREAD

**Issue by bkinsey808** [kumarharsh/graphql-for-vscode#1]
Does this project support the gql tag for graphql template literals?

**Comment by kumarharsh** [comment]
can you be more specific by "support"? do you mean syntax highlighting within the template literal?

**Comment by bkinsey808** [comment]
yes, correct.

**Comment by kumarharsh** [comment]
ok, thanks for clarifying. As far as I know, it's not possible as of this moment with vscode, as one of vscode's core members had stated in one of the issues elsewhere. I'll dig deeper and see what can be done about it. If possible, I'd also like to implement autocomplete for queries by tying it closely with the schemas.

**Comment by julienp** [comment]
I was poking a bit at this, and here's a proof of concept that enables the graphql (syntax highlighting) grammar inside Relay.QL and gql template strings: https://github.com/kumarharsh/graphql-for-vscode/compare/master...julienp:master

It also enables javascript syntax within ${} substitutions inside the template string.

**Comment by kumarharsh** [comment]
@julienp that sounds fantastic! I'll have a look tomorrow at this!

**Comment by kumarharsh** [comment]
@julienp I fail to see any noticeable change when trying out your changes. Is there any extra step you've done which are not in the commits you've pushed? I've tried this with vscode 1.7-insider and 1.8-insider.

For example, in the code from my master branch, I can see that the JS syntax within ${} is already enabled. Also, the graphql syntax highlighting doesn't seem to be working. See attached screenshots here:

Before:
![image](https://cloud.githubusercontent.com/assets/760112/20247922/0d5fa98a-a9ff-11e6-996b-0fc8a3132978.png)

After:
![image](https://cloud.githubusercontent.com/assets/760112/20247917/e5aaedb4-a9fe-11e6-8f13-22e1c3fc524c.png)

**Comment by julienp** [comment]
The problem is that fragmentName is not optional in the current grammar, but should be. See https://github.com/julienp/graphql-for-vscode/commit/b1689e8f04e21085b6dbb20b0314d844132620e3

Also there's an issue with the ${} substitution not being picked up inside a selection sets, but I haven't had time to track this down yet.

<img width="430" alt="screen shot 2016-11-13 at 20 59 27" src="https://cloud.githubusercontent.com/assets/387068/20248427/3f733ac4-a9e4-11e6-9140-42513d857111.png">

**Comment by kumarharsh** [comment]
@bkinsey808 @julienp I've added support for highlighting within template literals in the latest release. Have a look, and let me know if there are any bugs, etc. One thing to note is that syntax highlighting doesn't work when the language is set as "Javascript (Babel)",

**Comment by bkinsey808** [comment]
This is what I see now, looks pretty good, thanks.

![screenshot from 2016-11-28 23-25-59](https://cloud.githubusercontent.com/assets/475211/20704083/31e175da-b5c2-11e6-9e78-a6a216abc505.png)

**Comment by kumarharsh** [comment]
I'm going to close this issue now. If there are more issues, kindly open a new one.

**Comment by sparcut** [comment]
[Microsoft/vscode/issues/5961](https://github.com/Microsoft/vscode/issues/5961#issuecomment-335964652) - Support now avaliable! :tada:

**Comment by pungggi** [comment]
nice! will it now be possible to have intellisense in the query?

**Comment by sparcut** [comment]
@pungggi Theoretically, yes.

Microsoft's example of using this in an extension is [Microsoft/typescript-lit-html-plugin](https://github.com/Microsoft/typescript-lit-html-plugin)

**Comment by sparcut** [comment]
@kumarharsh Are you working on this, or should I make a PR?

**Comment by Mayank1791989** [comment]
@pungggi and @sparcut this plugin already support intellisense in queries from v1.1.0.

**Comment by Mayank1791989** [comment]
@ntziolis  plz share plugin output.

**Comment by ntziolis** [comment]
```
[Info  - 19:50:35] NODE_PATH value is: (path is good i checked)
[Info  - 19:50:36] Module '@playlyfe/gql' loaded from: .\node_modules\@playlyfe\gql\lib\index.js
[vscode] File extensions registered:  [ 'graphql', 'ts' ]
[Watch established (gqlSchemaFiles)] 
	basePath: (path that matches my project root) 
	RelativePath: undefined 
	Files: "generated/schema.graphql"
[Watch established (gqlQueryFiles-1)] 
	basePath: (path that matches my project root) 
	RelativePath: undefined 
	Files: "src/**/*.graphql"
[Watch established (gqlQueryFiles-0)] 
	basePath: (path that matches my project root) 
	RelativePath: undefined 
	Files: "src/**/*.ts"
```

**Comment by kumarharsh** [comment]
The syntax highlighting might notbe working due to interference with some other plugin. Can you disable some of them and check?

**Comment by Mayank1791989** [comment]
@ntziolis enable plugin debug mode  (`grahqlForVSCode.debug: true`) then restart vscode and share plugin output. Is autocomplete working in your query src/**/*.graphql files?

**Comment by ntziolis** [comment]
I took some time to setup a clean project + disabled all extension besides graphql for vscode. Here is my status and current issues:
- .graphql files
  - linting working for fragments, queries, mutations AND subscriptions
     - this means that the schema is properly wired up
  - typeahead working only after specifying either fragments, query or mutation manually
  - **incorrect typeahead for subscriptions**
- gql tag
  - linting working for fragments, queries, mutations AND subscriptions (same as for graphql files)
  - typeahead in empty file showing correct initial suggestions (fragments, queries, mutations AND subscriptions, this is different than for graphql files where wrong suggestions)
  - when going deeper typeahead continues to work for fragments, query and mutation (same as graphql files)
  - incorrect typeahead for subscriptions (same as graphql files)

### Config
```
{
  schema: {
    files: 'generated/schema.graphql',
    "validate": {
      "extends": "gql-rules-schema",
      "rules": {
        "NoUnusedTypeDefinition": "off"
      }
    }
  },
  query: {
     files: [
      {
        match: 'src/**/*.graphql',
        parser: 'QueryParser',
        "validate": { "extends": "gql-rules-query" }
      }, {
        match: 'src/**/*.ts',
        parser: ['EmbeddedQueryParser', { startTag: 'gql`', endTag: '`' }],
        "validate": { "extends": "gql-rules-query" }
      }
    ]
  }
}
```


### Log
With `"grahqlForVSCode.debug": true`
```
[Info  - 21:42:20] NODE_PATH value is: c:\vscode-graphql-plugin-test;C:\AppData\Roaming\npm
[Info  - 21:42:21] Module '@playlyfe/gql' loaded from: c:\vscode-graphql-plugin-test\node_modules\@playlyfe\gql\lib\index.js
[vscode] File extensions registered:  [ 'graphql', 'ts' ]
[Watch established (gqlSchemaFiles)] 
	basePath: C:/vscode-graphql-plugin-test 
	RelativePath: undefined 
	Files: "generated/schema.graphql"
[Watch established (gqlQueryFiles-0)] 
	basePath: C:/vscode-graphql-plugin-test 
	RelativePath: undefined 
	Files: "src/**/*.graphql"
[Watch established (gqlQueryFiles-1)] 
	basePath: C:/vscode-graphql-plugin-test 
	RelativePath: undefined 
	Files: "src/**/*.ts"
```
### Incorrect typeahead for subscriptions
Subscriptions however when manually specifying `subscription` I get the wrong typehead:
![image](https://user-images.githubusercontent.com/265338/33759957-6cf918b2-dc04-11e7-8a5f-e27d1164e7b3.png)

No error in output graphql console.

When I look at the typeahead it provides, it looks like the result I should be getting when the file / gql string is empty. This is works in neither `.graphql` files nor `gql` tag.

The same typeahead is provided no matter on which level in the subscription you are.

Please keep in mind that **linting works for subscriptions** so generally the schema seems to be wired up correctly, even for subscriptions.

### Typeahead working only after specifying either fragments, query or mutation manually
When I create a new graphql file the plugin doesn't suggest `query`, `mutation` or `subscription`:
![image](https://user-images.githubusercontent.com/265338/33759652-6860772e-dc03-11e7-956e-324301cfaca5.png)

In the output I can see that the following error is logged every time i invoke typeahead via ctrl+space:
```
[Error - 10:45:31] Request textDocument/completion failed.
  Message: Request textDocument/completion failed with message: expected style should have some value
  Code: -32603 
```
But once I manually specify `fragment`, `query` or `mutation` it provides correct typeahead and linting:
![image](https://user-images.githubusercontent.com/265338/33759774-cc8f14e4-dc03-11e7-9e67-a22363ffea49.png)

And there is no additional output in the graphql status console, neither error or anything else (which I assume is the desired state).

When doing the same in an empty gql tag I get the correct suggestions:
![image](https://user-images.githubusercontent.com/265338/33784909-f62849e2-dc62-11e7-8516-334f6736b98f.png)

No error output in the graphql console.

**Comment by Mayank1791989** [comment]
@ntziolis Looks like issue with gql tags is fixed. Did you figure out why it was not working previously?

> Typeahead working only after specifying either fragments, query or mutation manually

It's a bug in gql lib. I have created an issue for that https://github.com/Mayank1791989/gql/issues/78.

> Incorrect typeahead for subscriptions

gql currently does not properly support subscriptions see issue https://github.com/Mayank1791989/gql/issues/79

**Comment by kumarharsh** [comment]
Boom! 🎉 These issues are fixed in latest release. Please update your gql library to the latest version and give it a go.

**Comment by ntziolis** [comment]
@Mayank1791989 @kumarharsh Wow, didnt expect a 2 day turnaround. THANK YOU.

In regards to install issues, I will update if I stumble again, but to be honest initially it might have been that I was testing some of the non supported cases unaware they were not supported.

**Comment by FezVrasta** [comment]
the template literals highlighting isn't working for me, do I have to do anything special?

I have `@playlyfe/gql` installed globally, and the config set to`"graphqlForVSCode.nodePath": "~/.config/yarn/global/node_modules"`.

I also have the `.gqlconfig` file in my project folder (under `/apps/web/.gqlconfig`).

The `.gql` files are highlighted.

**Comment by kumarharsh** [comment]
Can you share a file which is not working? I'd like to take a look.

**Comment by FezVrasta** [comment]
any file, even ones with just

```
import gql from 'graphql-tag';
gql`
  query {
    name
  }
`
```

**Comment by kumarharsh** [comment]
I'm getting it highlighted properly:

![image](https://user-images.githubusercontent.com/760112/34255465-d20b6800-e676-11e7-88ef-0ef65d6b95e9.png)

What language is set for the file?
Can you also try to turn off some other plugins - I've been told that there are some plugins which interfere with this one's highlighting.

**Comment by FezVrasta** [comment]
I tried with "Javascript" and "Javascript React", I haven't any plugin I could simply turn off without lose anything I use daily, I think the extension should play nicely with the other extensions, otherwise it doesn't make much sense to me.

**Comment by kumarharsh** [comment]
Yes, I agree. I just want to get more information so that I can trace which part is conflicting. Can you disable all extensions, and then enable only this one, and do a "Reload Window"?

![image](https://user-images.githubusercontent.com/760112/34255639-9c299fa8-e677-11e7-8190-643e7f691e7d.png)

![image](https://user-images.githubusercontent.com/760112/34255656-b1a3377c-e677-11e7-84ac-916d42874a73.png)
![image](https://user-images.githubusercontent.com/760112/34255664-c3c1475a-e677-11e7-80c4-7d74454ba5c3.png)

See if you get the highlighting?

**Comment by FezVrasta** [comment]
I reinstalled and configured it, with all the extensions disabled, the minimal example works, but any other file in my project still misses the syntax highlight.

It looks like the syntax doesn't work if I provide the second argument to `graphql`:

```
import { graphql } from 'react-apollo';
import gql from 'graphql-tag';
graphql(
  gql`
    query {
      name
    }
  `,
  {}
);
```

**Comment by kumarharsh** [comment]
Conversation moved to https://github.com/kumarharsh/graphql-for-vscode/issues/62

**Comment by kumarharsh** [comment]
This extension does support gql tag. If you're not getting syntax highlighting, or have problems with autocomplete or go-to-definition, look through newer issues and add your voice there, or raise a new issue. This issue is very old, and discussion is tending to go off-track or notify older participants.
